src/compiler/checker.ts

```diff
@@ -21682,18 +21682,50 @@ namespace ts {
 
         function checkUnusedModuleMembers(node: ModuleDeclaration | SourceFile): void {
             if (compilerOptions.noUnusedLocals && !(node.flags & NodeFlags.Ambient)) {
+                const unusedImports = createMultiMap<ImportedDeclaration>();
                 node.locals.forEach(local => {
-                    if (!local.isReferenced && !local.exportSymbol) {
-                        for (const declaration of local.declarations) {
-                            if (!isAmbientModule(declaration)) {
-                                errorUnusedLocal(declaration, symbolName(local));
-                            }
+                    if (local.isReferenced || local.exportSymbol) return;
+                    for (const declaration of local.declarations) {
+                        if (isAmbientModule(declaration)) continue;
+                        if (isImportedDeclaration(declaration)) {
+                            unusedImports.add(String(getNodeId(importClauseFromImported(declaration))), declaration);
+                        }
+                        else {
+                            errorUnusedLocal(declaration, symbolName(local));
                         }
                     }
                 });
+
+                unusedImports.forEach(unuseds => {
+                    const importClause = importClauseFromImported(first(unuseds)); // others will have the same clause

```

Comment:
Why recompute this?  Isn't it the map key?

Reply:
Unfortunately we don't support non-string keys for maps for compatibility with ES5. See `shimMap`.

Reply:
We could always make the map value be a pair though...

Reply:
I must have been thinking of another context.  This makes sense, given that limitation.